iceoryx2-bb/posix/src/process_state.rs

```diff
@@ -0,0 +1,808 @@
+    ///     // resources
+    ///     ProcessState::Dead => (),
+    ///
+    ///     // The monitored process does not exist, maybe it did not yet start or already performed
+    ///     // a clean shutdown.
+    ///     ProcessState::DoesNotExist => (),
+    ///
+    ///     // The monitored process crashed and another process acquired the [`ProcessCleaner`]
+    ///     // to remove its remaining resources
+    ///     ProcessState::CleaningUp => (),
+    /// }
+    /// ```
+    pub fn state(&self) -> Result<ProcessState, ProcessMonitorStateError> {
+        let msg = "Unable to acquire ProcessState";
+        match unsafe { &*self.file.as_ptr() } {
+            Some(_) => self.read_state_from_file(),
+            None => match File::does_exist(&self.path) {
+                Ok(true) => {
+                    self.file.set(Self::open_file(&self.path)?);
+                    self.read_state_from_file()
+                }
+                Ok(false) => Ok(ProcessState::DoesNotExist),
+                Err(v) => {
+                    fail!(from self, with ProcessMonitorStateError::UnknownError(0),
+                        "{} since an unknown failure occurred while checking if the file exists ({:?}).", msg, v);
+                }
+            },
+        }
+    }
+
+    fn get_lock_state(file: &File) -> Result<i64, ProcessMonitorStateError> {
+        let msg = format!("Unable to acquire lock on file {:?}", file);
+        let mut current_state = posix::flock::new();
+        current_state.l_type = LockType::Write as _;
+
+        if unsafe {
+            posix::fcntl(
+                file.file_descriptor().native_handle(),
+                posix::F_GETLK,
+                &mut current_state,
+            )
+        } == -1
+        {
+            handle_errno!(ProcessMonitorStateError, from "ProcessMonitor::get_lock_state()",
+                Errno::EINTR => (Interrupt, "{} since an interrupt signal was received.", msg),
+                v => (UnknownError(v as i32), "{} since an unknown error occurred ({}).", msg, v)
+            )
+        }
+
+        Ok(current_state.l_type as _)
+    }
+
+    fn read_state_from_file(&self) -> Result<ProcessState, ProcessMonitorStateError> {
+        let file = match unsafe { &*self.file.as_ptr() } {
+            Some(ref f) => f,
+            None => return Ok(ProcessState::Starting),
+        };
+
+        let msg = format!("Unable to read state from file {:?}", file);
+        let lock_state = fail!(from self, when Self::get_lock_state(file),
+                            "{} since the lock state of the state file could not be acquired.", msg);
+
+        match lock_state as _ {
+            posix::F_WRLCK => Ok(ProcessState::Alive),
+            _ => match File::does_exist(&self.path) {
+                Ok(true) => match file.permission() {
+                    Ok(INIT_PERMISSION) => Ok(ProcessState::Starting),
+                    Err(_) | Ok(_) => {
+                        self.file.set(None);
```

Comment:
Is this correct? Setting it to `None` although no cleanup has taken place? So the one who calls `state` has to do the cleanup immediately and cannot rely on calling `state` later on to get the same result?

Reply:
It works. The `file` is required to check the lock state. If the lock state indicates that it is dead, we proactively close the file and reopen it again with every new run.

If we wouldn't do it, a `ProcessCleaner` would remove all resources but the monitor would still hold the old file descriptor pointing to a file that no longer exists in the file system.

We have no other choice than to close it and reopen it with every run to not run into this problem.